urlc175f05cb8_mchllbrt_PermCode_tgz-pJ8-permlib_examples_UniversalSeparableJ8/src/permlib/examples/RandomPermutationClass.java

```diff
@@ -0,0 +1,183 @@
+//        }
+//        System.out.println();
+//        for (Permutation p : ur(7)) {
+//            System.out.println(p);
+//        }
+//        System.out.println();
+
+//        for(Permutation p : io(7)) System.out.println(p);
+    }
+
+    /* In Av(321) randomly choose a new basis element of each length */
+    public static void example01(int k) {
+        long s = System.currentTimeMillis();
+        // System.out.println("Seed: " + s);
+        R.setSeed(s);
+        ArrayList<Permutation> basis = new ArrayList<>();
+        basis.add(new Permutation("321"));
+        for (int n = k; n < 15; n++) {
+            int count = 0;
+            PermutationClass c = new PermutationClass(basis);
+            Permutation newBasis = new Permutation(n);
+            boolean hasSimple = false;
+            boolean hasIO = true;
+            for (Permutation p : io(n)) {
+                hasIO &= c.containsPermutation(p);
+            }
+            if (!hasIO) {
+                System.out.println("IO fail");
+                return;
+            }
+            if (n >= 5) {
+                boolean hasUR = false;
+                for (Permutation p : ur(n)) {
+                    hasUR |= c.containsPermutation(p);
+                }
+                if (!hasUR) {
+                    System.out.println("UR fail");
+                    return;
+                }
+                boolean hasUL = false;
+                for (Permutation p : ul(n)) {
+                    hasUL |= c.containsPermutation(p);
+                }
+                if (!hasUL) {
+                    System.out.println("UL fail");
+                    return;
+                }
+            }
+            for (Permutation p : new Permutations(c, n)) {
+                hasSimple |= PermUtilities.SIMPLE.isSatisfiedBy(p);
+                count++;
+                if (R.nextInt(count) == 0) {
+                    newBasis = p;
+                }
+            }
+           // System.out.println(newBasis + " " + count);
+            if (!hasSimple) {
+                System.out.println("Simple fail");
+                return;
+            }
+            if (newBasis.equals(new Permutation(n))) {
+                System.out.println("Monotone fail");
+                return;
+            }
+            basis.add(newBasis);
+        }
+        System.out.print("Passed ");
+        for(Permutation p : basis) System.out.print(p + ", ");
+        System.out.println();
+    }
+
+    public static Permutation[] io(int n) {
+        if (n == 4) {
+            return new Permutation[]{new Permutation("2413"), new Permutation("3142")};
+        }
+        Permutation[] result = io(n - 1);
+        if (n % 2 == 0) {
+            result[0] = result[0].insert(n - 1, n - 2);
+            result[1] = result[1].insert(n - 2, n - 1);
+        } else {
+            result[1] = result[1].insert(n - 1, n - 2);
+            result[0] = result[0].insert(n - 2, n - 1);
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 116:122, 135:141.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.